java/src/main/java/com/saucelabs/simplesauce/SauceOptions.java

```diff
@@ -1,134 +1,150 @@
+                        continue;
+                    } else if (isInEnum(key, Options.W3C.class)) {
+                        w3cCapabilities.setCapability(fieldName, value);
+                    } else if (isInEnum(key, Options.SAUCE.class)) {
+                        sauceCapabilities.setCapability(fieldName, value);
+                    } else if ("publicRestricted".equals(key)) {
+                        sauceCapabilities.setCapability("public", value);
+                    }
+                }
+            } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+                e.printStackTrace();
+            }
+        }
+
+        w3cCapabilities.setCapability("sauce:options", sauceCapabilities);
+        return w3cCapabilities;
+    }
+
+    public String getBuild() {
+        if (build != null) {
+            return build;
+            // Jenkins
+        } else if (System.getenv("BUILD_TAG") != null) {
+            return System.getenv("BUILD_NAME") + ": " + System.getenv("BUILD_NUMBER");
+            // Bamboo
+        } else if (System.getenv("bamboo_agentId") != null) {
+            return System.getenv("bamboo_shortJobName") + ": " + System.getenv("bamboo_buildNumber");
+            // Travis
+        } else if (System.getenv("TRAVIS_JOB_ID") != null) {
+            return System.getenv("TRAVIS_JOB_NAME") + ": " + System.getenv("TRAVIS_JOB_NUMBER");
+            // CircleCI
+        } else if (System.getenv("CIRCLE_JOB") != null) {
+            return System.getenv("CIRCLE_JOB") + ": " + System.getenv("CIRCLE_BUILD_NUM");
+            // Gitlab
+        } else if (System.getenv("CI") != null) {
+            return System.getenv("CI_JOB_NAME") + ": " + System.getenv("CI_JOB_ID");
+            // Team City
+        } else if (System.getenv("TEAMCITY_PROJECT_NAME") != null) {
+            return System.getenv("TEAMCITY_PROJECT_NAME") + ": " + System.getenv("BUILD_NUMBER");
+            // Default
+        } else {
+            return "Build Time: " + Long.toString(System.currentTimeMillis());
+        }
+    }
+
+    public void setCapabilities(Map<String, Object> capabilities) {
+        capabilities.forEach(this::setCapability);
+    }
+
+    // TODO: Consider making this public
+    private void setCapability(String key, Object value) {
+        String setter = "set" + key.substring(0, 1).toUpperCase() + key.substring(1);
+        try {
+            Class<?> type = SauceOptions.class.getDeclaredField(key).getType();
+            Method method = SauceOptions.class.getDeclaredMethod(setter, type);
+            method.invoke(this, value);
+        } catch (NoSuchMethodException | NoSuchFieldException | InvocationTargetException | IllegalAccessException e) {
+            e.printStackTrace();
+        }
+    }
+
+    // TODO: Consider making this public
+    private Object getCapability(String capability) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
```

Comment:
Prefer catching and handling the exceptions in this method, rather than leaking them out if you're thinking of making this `public`